### PR TITLE
add `async` walredo mode (disabled-by-default, opt-in via config)

### PR DESCRIPTION
**Pull request by problame**
Before this PR, the `nix::poll::poll` call would stall the executor.

This PR refactors the `walredo::process` module to allow for different implementations, and adds a new `async` implementation which uses `tokio::process::ChildStd{in,out}` for IPC.

The `sync` variant remains the default for now; we'll do more testing in staging and gradual rollout to prod using the config variable.

Performance
-----------

I updated `bench_walredo.rs`, demonstrating that a single `async`-based walredo manager used by N=1...128 tokio tasks has lower latency and higher throughput.

I further did manual less-micro-benchmarking in the real pageserver binary.
Methodology & results are published here:
https://neondatabase.notion.site/2024-04-08-async-walredo-benchmarking-8c0ed3cc8d364a44937c4cb50b6d7019?pvs=4

tl;dr:
- use pagebench against a pageserver patched to answer getpage request & small-enough working set to fit into PS PageCache / kernel page cache.
- compare knee in the latency/throughput curve
    - N tenants, each 1 pagebench clients
    - sync better throughput at N < 30, async better at higher N
    - async generally noticable but not much worse p99.X tail latencies
    - eyeballing CPU efficiency in htop, `async` seems significantly more CPU efficient at ca N=[0.5*ncpus, 1.5*ncpus], worse than `sync` outside of that band

Mental Model For Walredo & Scheduler Interactions
-------------------------------------------------

Walredo is CPU-/DRAM-only work.
This means that as soon as the Pageserver writes to the pipe, the walredo process becomes runnable.

To the Linux kernel scheduler, the `$ncpus` executor threads and the walredo process thread are just `struct task_struct`, and it will divide CPU time fairly among them.

In `sync` mode, there are always `$ncpus` runnable `struct task_struct` because the executor thread blocks while `walredo` runs, and the executor thread becomes runnable when the `walredo` process is done handling the request.
In `async` mode, the executor threads remain runnable unless there are no more runnable tokio tasks, which is unlikely in a production pageserver.

The above means that in `sync` mode, there is an implicit concurrency limit on concurrent walredo requests (`$num_runtimes * $num_executor_threads_per_runtime`).
And executor threads do not compete in the Linux kernel scheduler for CPU time, due to the blocked-runnable-ping-pong.
In `async` mode, there is no concurrency limit, and the walredo tasks compete with the executor threads for CPU time in the kernel scheduler.

If we're not CPU-bound, `async` has a pipelining and hence throughput advantage over `sync` because one executor thread can continue processing requests while a walredo request is in flight.

If we're CPU-bound, under a fair CPU scheduler, the *fixed* number of executor threads has to share CPU time with the aggregate of walredo processes.
It's trivial to reason about this in `sync` mode due to the blocked-runnable-ping-pong.
In `async` mode, at 100% CPU, the system arrives at some (potentially sub-optiomal) equilibrium where the executor threads get just enough CPU time to fill up the remaining CPU time with runnable walredo process.

Why `async` mode Doesn't Limit Walredo Concurrency
--------------------------------------------------

To control that equilibrium in `async` mode, one may add a tokio semaphore to limit the number of in-flight walredo requests.
However, the placement of such a semaphore is non-trivial because it means that tasks queuing up behind it hold on to their request-scoped allocations.
In the case of walredo, that might be the entire reconstruct data.
We don't limit the number of total inflight Timeline::get (we only throttle admission).
So, that queue might lead to an OOM.

The alternative is to acquire the semaphore permit *before* collecting reconstruct data.
However, what if we need to on-demand download?

A combination of semaphores might help: one for reconstruct data, one for walredo.
The reconstruct data semaphore permit is dropped after acquiring the walredo semaphore permit.
This scheme effectively enables both a limit on in-flight reconstruct data and walredo concurrency.

However, sizing the amount of permits for the semaphores is tricky:
- Reconstruct data retrieval is a mix of disk IO and CPU work.
- If we need to do on-demand downloads, it's network IO + disk IO + CPU work.
- At this time, we have no good data on how the wall clock time is distributed.

It turns out that, in my benchmarking, the system worked fine without a semaphore. So, we're shipping async walredo without one for now.

Future Work
-----------

We will do more testing of `async` mode and gradual rollout to prod using the config flag.
Once that is done, we'll remove `sync` mode to avoid the temporary code duplication introduced by this PR.
The flag will be removed.

The `wait()` for the child process to exit is still synchronous; the comment [here]( https://github.com/neondatabase/neon/blob/655d3b64681b6562530665c9ab5f2f806f30ad01/pageserver/src/walredo.rs#L294-L306) is still a valid argument in favor of that.

The `sync` mode had another implicit advantage: from tokio's perspective, the calling task was using up coop budget.
But with `async` mode, that's no longer the case -- to tokio, the writes to the child process pipe look like IO.
We could/should inform tokio about the CPU time budget consumed by the task to achieve fairness similar to `sync`.
However, the [runtime function for this is `tokio_unstable`](`https://docs.rs/tokio/latest/tokio/task/fn.consume_budget.html).

Refs
----

refs #6628 
refs https://github.com/neondatabase/neon/issues/2975

### CODE REVIEW

**Review by koivunej — CHANGES_REQUESTED** [review]
We should add to `output.pending_responses` a way to mark that this response will never be read for cancelled requests... Sadly, this is important for those requests which are waiting on `self.stdout.lock().await` so they cannot just "push" something there.

Alternatives are fine, because my http testing endpoint is probably only one which could be dropped right now. Alternatives such as "log an error from scopeguard".

Similarly it would be really bad to have the stdin writing be cancelled midway.

**Comment by problame** [comment]
Good catch with the cancellation problem.

Pushed changes, please have another look.

File: libs/utils/src/poison.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                unreachable!("we fail check_and_arm() if it's in that state: {at:?}")
+            }
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("poisoned at {at:?}: {what}")]
+    Poisoned { what: &'static str, at: Instant },
```
Comment (koivunej) [inline libs/utils/src/poison.rs:119]: This `at` will likely be relative to the host last boot, perhaps a `chrono::DateTime<chrono::Utc>` instead?

**Review by koivunej — APPROVED** [review]

**Review by knizhnik — APPROVED** [review]

**Comment by problame** [comment]
Joonas has found performance of this PR to be much worse, DM : https://neondb.slack.com/archives/D049K7HJ9JM/p1707009731704869

```
short/short/1           time:   [11.417 µs 11.635 µs 11.846 µs]
Found 18 outliers among 100 measurements (18.00%)
  4 (4.00%) low severe
  5 (5.00%) low mild
  9 (9.00%) high mild
short/short/2           time:   [33.945 µs 34.371 µs 34.795 µs]
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  4 (4.00%) low mild
  2 (2.00%) high mild
short/short/4           time:   [55.483 µs 55.901 µs 56.294 µs]
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) low severe
  1 (1.00%) low mild
  5 (5.00%) high mild
short/short/8           time:   [126.11 µs 126.67 µs 127.22 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
short/short/16          time:   [257.16 µs 258.27 µs 259.40 µs]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

medium/medium/1         time:   [76.750 µs 78.309 µs 79.861 µs]
medium/medium/2         time:   [169.51 µs 170.02 µs 170.61 µs]
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  7 (7.00%) high severe
medium/medium/4         time:   [305.85 µs 308.94 µs 312.44 µs]
Found 10 outliers among 100 measurements (10.00%)
  8 (8.00%) high mild
  2 (2.00%) high severe
medium/medium/8         time:   [555.67 µs 560.14 µs 564.97 µs]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
medium/medium/16        time:   [1.1525 ms 1.1644 ms 1.1773 ms]
Found 9 outliers among 100 measurements (9.00%)
  7 (7.00%) high mild
  2 (2.00%) high severe
```

>  these are very much inline with what I remember, but yeah perhaps the single thread version used to be below 10us, maybe 7us. my shmempipe got to 2-3us for the short, which I think is the limit I guess. that was with probably way too much spinning :) (edited)

**Comment by problame** [comment]
I pushed https://github.com/neondatabase/neon/pull/6548/commits/cca66e5e82ba8657a03e1a05d38c180fd5c3d7e7  plus https://github.com/neondatabase/neon/pull/6548/commits/67a7abc7cf4a377bbe0d3283a498f6851538b1f3 which make the walredo process implementation configurable.

There's too much code duplication for my taste, but, it was the easiest way to have the implementations side by side.

The runtime-override is a nice bonus & makes benchmarking simpler, so, I would want to keep it until we commit to clearly one vs the other walredo variant.

I'm not sure whether we should actually have the side-by-side support. My [recent benchmarking](https://github.com/neondatabase/neon/issues/6628#issuecomment-2025015263) suggests that async walredo is better at throughput and marginally worse at tail latencies.

So,  I'll possibly revert these two commits.

**Comment by problame** [comment]
@koivunej the PR changed significantly since your last review, re-requesting it. I think we have established context in various meetings / calls in the last couple of weeks.

File: pageserver/src/walredo/process/process_impl/process_async.rs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+            .id()
+    }
+
+    /// Apply given WAL records ('records') over an old page image. Returns
+    /// new page image.
+    ///
+    /// # Cancel-Safety
+    ///
+    /// Cancellation safe.
+    #[instrument(skip_all, level = tracing::Level::DEBUG, fields(tenant_id=%self.tenant_shard_id.tenant_id, shard_id=%self.tenant_shard_id.shard_slug(), pid=%self.id()))]
```
Comment (koivunej) [inline pageserver/src/walredo/process/process_impl/process_async.rs:178]: I wonder if the `tenant_id`, `shard_id` are needed, I think in all uses we should have those already. `pid` is useful.

File: pageserver/src/walredo/process.rs [hunk 3]
```diff
@@ -1,408 +1,136 @@
...
+impl TryFrom<u8> for Kind {
+    type Error = u8;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        Ok(match value {
+            v if v == (Kind::Sync as u8) => Kind::Sync,
+            v if v == (Kind::Async as u8) => Kind::Async,
+            x => return Err(x),
+        })
+    }
```
Comment (koivunej) [inline pageserver/src/walredo/process.rs:46]: I do not understand when is this needed?

Comment (koivunej) [reply pageserver/src/walredo/process.rs:46]: It seems to be dead code. No dead code warnings even with `pub(crate)` though.

File: pageserver/src/walredo/process.rs [hunk 4]
```diff
@@ -1,408 +1,136 @@
...
 }
 
-struct ProcessOutput {
-    stdout: ChildStdout,
-    pending_responses: VecDeque<Option<Bytes>>,
-    n_processed_responses: usize,
+impl Kind {
+    // TODO: remove
+    pub const DEFAULT: Kind = Kind::Sync;
+    pub const DEFAULT_TOML: &'static str = "sync";
```
Comment (koivunej) [inline pageserver/src/walredo/process.rs:52]: Interestingly if `Kind` is `pub(crate)` then `DEFAULT` is unused. Strum cannot be used with the const as the `IntoStaticStr` only generates a `impl From<Kind> to &'static str` not an inherent method as I'd manually do.

Were you going to delete the DEFAULT anyways before merging?

File: pageserver/src/walredo/process/process_impl/process_async.rs (hunk 2 above)
Comment (problame) [reply pageserver/src/walredo/process/process_impl/process_async.rs:178]: It's pre-existing code / copy-pasted over from process_sync. Let's leave it as is, can be cleaned up once we remove `process_sync` after rolling out async to the fleet.

Comment (problame) [reply pageserver/src/walredo/process/process_impl/process_async.rs:178]: Added a todo to the epic to keep track of this

File: pageserver/src/walredo/process/process_impl/process_std.rs [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+        }};
+    }
+        set_nonblock_or_log_err!(stdin)?;
+        set_nonblock_or_log_err!(stdout)?;
+
+        // all fallible operations post-spawn are complete, so get rid of the guard
+        let child = scopeguard::ScopeGuard::into_inner(child);
+
+        tokio::spawn(
+        async move {
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/walredo/process/process_impl/process_std.rs:116]: > if we would pull out a `let span = ...`

Does not help.

File: pageserver/src/walredo/process.rs (hunk 4 above)
Comment (problame) [reply pageserver/src/walredo/process.rs:52]: Simplified this in 005dcbd6a89f06db2577edfb51d3aea0f287d491

Comment (koivunej) [reply pageserver/src/walredo/process.rs]: Ok I missed the test uses it. `#[cfg(test)]` would clarify.

**Review by koivunej — APPROVED** [review]
Approving again, diff looks reasonable.
